### ISSUE TITLE
[Rust] Potential memory leak issue

### ISSUE THREAD

**Issue by hydai** [WasmEdge/WasmEdge#1805]
## Description

Users report this issue.
The reproduced repo: https://github.com/wujunzhuo/wasm-test

Related problematic codes:

```rust
extern "C" {
    fn load_input(pointer: *mut u8);
    fn dump_output(pointer: *const u8, length: usize);
}

#[no_mangle]
pub extern "C" fn handler(input_length: usize) {
    // load input data
    let mut input = Vec::with_capacity(input_length);
    unsafe {
        load_input(input.as_mut_ptr());
        input.set_len(input_length);
    }

    // process app data
    let output = input.to_ascii_uppercase();

    // dump output data
    unsafe {
        dump_output(output.as_ptr(), output.len());
    }
}
```

### Current State

The same handler function written in C(via clang + wasi-libc) and Golang(via tinaygo) works without any issue. However, the memory will increase by 16KB every time when calling the handler function several times when using the wasm compiled from the rust version. You can check the behavior via ps.

### Expected

If the rust code is correct, the memory should not increase every time.

## Environment

- Hardware Architecture: x86_64
- Operating system: Ubuntu 20.04

**Comment by wujunzhuo** [comment]
Update more details on 2022.9.1

I simplified my rust code to only one guest function:

*src/lib.rs*
```rust
#[no_mangle]
pub extern "C" fn hello() {
    println!("hello");
}
```
compile with `cargo build --release --target wasm32-wasi` (cargo 1.63.0)

Then I use the wasmedge-go sdk as runtime to call the `hello` function periodically:

*main.go*
```go
package main

import (
	"os"
	"time"

	"github.com/second-state/WasmEdge-go/wasmedge"
)

func hello(_ any, _ *wasmedge.CallingFrame, _ []any) ([]any, wasmedge.Result) {
	return nil, wasmedge.Result_Success
}

func main() {
	conf := wasmedge.NewConfigure(wasmedge.WASI)
	defer conf.Release()
	vm := wasmedge.NewVMWithConfig(conf)
	defer vm.Release()

	wasi := vm.GetImportModule(wasmedge.WASI)
	wasi.InitWasi(
		nil,
		os.Environ(),
		[]string{".:."},
	)

	module := wasmedge.NewModule("env")
	defer module.Release()
	f := wasmedge.NewFunction(wasmedge.NewFunctionType(
		[]wasmedge.ValType{},
		[]wasmedge.ValType{}), hello, nil, 0)
	module.AddFunction("hello", f)
	vm.RegisterModule(module)

	vm.LoadWasmFile("target/wasm32-wasi/release/hello.wasm")
	vm.Validate()
	vm.Instantiate()

	for {
		time.Sleep(100 * time.Millisecond)
		vm.Execute("hello")
	}
}
```

The following shell script shows that the memory usage keeps increasing:

*show_mem.sh*
```sh
#!/bin/bash

while true
do
    sleep 3
    date
    ps -p $(pidof hello) -o %mem,rss
done
```

```sh
root@iuiajpa5c06qjchk:~/tmp/wasm# ./show_mem.sh
Thu Sep  1 05:38:06 AM UTC 2022
%MEM   RSS
 2.6 26648
Thu Sep  1 05:38:09 AM UTC 2022
%MEM   RSS
 2.6 26712
Thu Sep  1 05:38:12 AM UTC 2022
%MEM   RSS
 2.6 26780
Thu Sep  1 05:38:15 AM UTC 2022
%MEM   RSS
 2.7 26844
Thu Sep  1 05:38:18 AM UTC 2022
%MEM   RSS
 2.7 26908
Thu Sep  1 05:38:21 AM UTC 2022
%MEM   RSS
 2.7 26984
^C
```

However, if I comment the `println` line in the rust file (leaves just an empty function), it seems no memory leak happens.

```sh
root@iuiajpa5c06qjchk:~/tmp/wasm# ./show_mem.sh 
Thu Sep  1 05:36:22 AM UTC 2022
%MEM   RSS
 2.4 24596
Thu Sep  1 05:36:25 AM UTC 2022
%MEM   RSS
 2.4 24596
Thu Sep  1 05:36:28 AM UTC 2022
%MEM   RSS
 2.4 24596
Thu Sep  1 05:36:31 AM UTC 2022
%MEM   RSS
 2.4 24596
Thu Sep  1 05:36:34 AM UTC 2022
%MEM   RSS
 2.4 24596
^C
```

*os*
```sh
root@iuiajpa5c06qjchk:~/tmp/wasm# uname -a
Linux iuiajpa5c06qjchk 5.15.0-47-generic #51-Ubuntu SMP Thu Aug 11 07:51:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

Thanks~

**Comment by apepkuss** [comment]
Hi @hydai @wujunzhuo ,

As a reference, I used Rust SDK to try the same test with `WasmEdge 0.11.0` + `wasmedge-sdk 0.4.0` on MacOS + M1Pro. The `hello.wasm` is same, and the main function is defined as below:

```rust
use wasmedge_sdk::{
    config::{CommonConfigOptions, ConfigBuilder, HostRegistrationConfigOptions},
    error::HostFuncError,
    params, CallingFrame, ImportObjectBuilder, Module, Vm, WasmValue,
};

fn hello(_: &CallingFrame, _: Vec<WasmValue>) -> Result<Vec<WasmValue>, HostFuncError> {
    println!("Hello, world!");

    Ok(vec![])
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let config = ConfigBuilder::new(CommonConfigOptions::default())
        .with_host_registration_config(HostRegistrationConfigOptions::new().wasi(true))
        .build()?;

    // create a Vm context
    let mut vm = Vm::new(Some(config))?;

    // get default wasi module
    let mut wasi_module = vm.wasi_module()?;
    let vec_envs: Vec<String> = std::env::vars()
        .into_iter()
        .map(|(k, v)| format!("{}={}", k, v))
        .collect();
    let envs = vec_envs.iter().map(|s| s.as_str()).collect();
    let preopens = vec![".:."];
    wasi_module.initialize(None, Some(envs), Some(preopens));

    let import = ImportObjectBuilder::new()
        .with_func::<(), ()>("hello", hello)?
        .build("env")?;
    let vm = vm.register_import_module(import)?;

    let module = Module::from_file(
        None,
        "/Users/sam/workspace/rust/mem-test/hello/target/wasm32-wasi/release/hello.wasm",
    )?;
    let vm = vm.register_module(None, module)?;

    let mut round = 1;
    loop {
        println!("Round_{}", round);
        vm.run_func(None, "hello", params![])?;
        round += 1;
    }

    Ok(())
}
```

The following two diagrams are rendered based on the 610 records generated by `show_mem.sh` in a test running approx. 30 min. The first one is presenting the memory consumption by percentage, while the second is the RSS data. Hope the work can help you!

<img width="532" alt="image" src="https://user-images.githubusercontent.com/4726889/187930988-4f17ae36-3571-4892-973f-f9621f46e75a.png">

<img width="531" alt="image" src="https://user-images.githubusercontent.com/4726889/187933968-f00fc98d-5ffb-4359-a034-a5629cb107dc.png">

**Comment by wujunzhuo** [comment]
@apepkuss Thanks for your response~ 👏

So what is the conclusion? Shall we see it as a bug of cargo-wasi and report to https://github.com/rust-lang/cargo?

**Comment by apepkuss** [comment]
@wujunzhuo I'd like to run the same test code on Ubuntu 20.04 + x86_64 so that we can check if the same test could perform differently in memory consumption. I'll try to do it asap and provide the data here. After that, we could draw a final conclusion. Thanks!

**Comment by apepkuss** [comment]
@hydai @wujunzhuo 

The following two diagrams show the memory consumption and RSS (in kb) on Ubuntu 20.04 (WSL2) + x86_64 with the exactly same test code as used on MacOS + M1Pro. Compared with the result on MacOS + M1Pro, the memory consumption on Ubuntu 20.04 + x86_64 reached the peak at the early stage, but did not fall down to a lower (and seems more reasonable) level.  It seems that some memory resources are not freed correctly. @hydai Does our runtime have different memory model design for different platforms?  

<img width="532" alt="image" src="https://user-images.githubusercontent.com/4726889/188079177-65adc846-21f7-416b-acd7-5a89c53d5fb6.png">

<img width="532" alt="image" src="https://user-images.githubusercontent.com/4726889/188079200-6530eecf-0a92-42c3-a415-6f4fea17b8e1.png">

**Comment by apepkuss** [comment]
@paulmchen Could you please share your test code with us if you don't mind? Thanks a lot!

**Comment by apepkuss** [comment]
Updated: 2023-04-06

We did the same test on `WasmEdge-0.12.0-alpha.2-38-gdad018af` with both Rust SDK and Go SDK. You can find the result [here](https://github.com/apepkuss/mem-test).

**Comment by wujunzhuo** [comment]
Confirmed that this is a bug from the Rust wasm toolchain `wasi-libc`, and has been fixed in Rust v1.69~

**Comment by dannypsnl** [comment]
Seems complete, close.
